Implement "Purchase Domain" screen UI

justtwago opened the pull request:
Fixes #19067

This PR contains the implementation of a "Purchase Domain" screen's UI. I also created a ViewModel setup that already handles all the screen's action buttons. To make it look a bit nicer in a landscape orientation, I slightly modified the UI layout so you can play with it.
Note: I added new analytics events, but they seem to be not registered, so I'll continue their setup either within this PR or in the next one when attaching the screen to a complete flow.

<details>
  <summary>Demo 🎥</summary>

https://github.com/wordpress-mobile/WordPress-Android/assets/16563318/18f50612-5c3a-48a5-8edd-2f1e56accecb

</details>

To test:
- [ ] Go to `PurchaseDomainScreen.kt` and open the preview panel.
- [ ] Choose any preview and hit on  <img width="24" alt="Screenshot 2023-10-06 at 4 36 37 PM" src="https://github.com/wordpress-mobile/WordPress-Android/assets/16563318/b16a3413-d1b1-4b3e-adcf-351a868956ad"> `Run Preview` button on the top right corner of the preview. The preview should be launched on your device or emulator.
- [ ] Play with the screen: change orientation, switch Light/Dark modes, and try different devices.

## Regression Notes
1. Potential unintended areas of impact
N/A (The composables are not yet in use, except for in previews, i.e. no app usage yet).

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

Commit:
Implement "Use Domain" screen UI

Commit:
Implement and test UseDomainViewModel and handle action buttons

Commit:
Fix MainTopAppBar javadoc and UseDomainViewModel.mainDispatcher access modifier

antonis commented on libs/analytics/src/main/java/org/wordpress/android/analytics/AnalyticsTracker.java:
```diff
@@ -1068,6 +1068,9 @@ public enum Stat {
         PERSONALIZATION_SCREEN_CARD_SHOW_TAPPED,
+        DOMAIN_MANAGEMENT_USE_DOMAIN_SCREEN_SHOWN,
+        DOMAIN_MANAGEMENT_USE_DOMAIN_SCREEN_NEW_DOMAIN_TAPPED,
+        DOMAIN_MANAGEMENT_USE_DOMAIN_SCREEN_EXISTING_DOMAIN_TAPPED
```
Thank you for adding the tracking events @justtwago 🙇 
I think those should also be added [at the end of the `AnalyticsTrackerNosara`](https://github.com/wordpress-mobile/WordPress-Android/blob/a4fa6ac15e45c3b616dc86d09529584fdd08c971/libs/analytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java#L2618):
```
            case DOMAIN_MANAGEMENT_USE_DOMAIN_SCREEN_SHOWN:
                return "domain_management_use_domain_screen_shown";
            case DOMAIN_MANAGEMENT_USE_DOMAIN_SCREEN_NEW_DOMAIN_TAPPED:
                return "domain_management_use_domain_screen_new_domain_tapped";
            case DOMAIN_MANAGEMENT_USE_DOMAIN_SCREEN_EXISTING_DOMAIN_TAPPED:
                return "domain_management_use_domain_screen_existing_domain_tapped";
```

mkevins replied:
Good catch Antonis! Maybe out of scope for this task (and certainly this PR), but I wonder whether these tracker files are good candidates for conversion to Kotlin 🤔 . This isn't the first time that stat cases have been missed.

antonis replied:
> I wonder whether these tracker files are good candidates for conversion to Kotlin 🤔 . This isn't the first time that stat cases have been missed.

Good idea @mkevins 👍 Kotlin would help on catching the missing cases and [that file might not be that complicated to convert](https://github.com/wordpress-mobile/WordPress-Android/blob/a4fa6ac15e45c3b616dc86d09529584fdd08c971/libs/analytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java).

justtwago replied:
Thank you! Good catch! I'd ticket the `AnalyticsTrackerNosara` conversion to Kotlin but I believe it shouldn't be placed on the `Domain Management` project's board. Where would you put it, folks?

antonis replied:
> Where would you put it, folks?

I don't think we have a project for this but we could add it under [the tech debt label](https://github.com/wordpress-mobile/WordPress-Android/labels/Tech%20Debt) 🤔

antonis reviewed (approved):
Hey @justtwago 👋 
I took the liberty of jumping to review this PR as I still introduce myself in the project.
I tested the implementation on a Pixel 5 (Android 14) and everything worked as expected for me in all configurations. The code changes also look good (though I'm a Compose newbie 😅) and the added tests are nice 🎉 

> To make it look a bit nicer in a landscape orientation, I slightly modified the UI layout so you can play with it.

The landscape behavior looks nice to me 🤩 

|Initial state|Scrolled state|
|---|---|
|![Screenshot_20231009_131556](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/c186c9f9-d859-4c77-8445-0a1804e14326)|![Screenshot_20231009_131612](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/c8306bb2-e7ea-4340-a479-6d395af75853)|

Looping in @paulforgione for design feedback if needed.

> Note: I added new analytics events, but they seem to be not registered, so I'll continue their setup either within this PR or in the next one when attaching the screen to a complete flow.

I've left a minor note on some missing boilerplate for the new track events (https://github.com/wordpress-mobile/WordPress-Android/pull/19332/files#r1350096183) but this is probably not blocking given that the implementation is not yet wired up to the app.

paulforgione commented:
Looking good! 

_The only potential change would be removing the line break from the title when it's being displayed in landscape mode._

If that's going to be a lot of work though, I'm happy to leave it as it (I don't imagine very many people are going to be navigating in landscape to begin with).

mkevins commented on WordPress/src/main/java/org/wordpress/android/ui/domains/management/usedomain/composable/UseDomainScreen.kt:
```diff
@@ -0,0 +1,249 @@
+@Preview(name = "Light mode", locale = "en")
+@Preview(name = "Dark mode", uiMode = Configuration.UI_MODE_NIGHT_YES)
+@Preview(name = "Small screen", device = Devices.NEXUS_5)
+@Preview(name = "Landscape orientation", device = Devices.AUTOMOTIVE_1024p)
```
Friends don't let friends WordPress and drive 😆 . Seriously, though, thanks for thinking about landscape!

mkevins commented on WordPress/src/main/java/org/wordpress/android/ui/domains/management/usedomain/composable/UseDomainScreen.kt:
```diff
@@ -0,0 +1,249 @@
+                Spacer(modifier = Modifier.width(16.dp))
+                ExistingDomainCard(onExistingDomainCardSelected, modifier = Modifier.weight(1f))
+            }
+        }
```
I think this is fine as it is, but just wanted to note that we could also achieve the necessary spacing via the `Row`'s and `Column`'s respective parameters `horizontalArrangement` and `verticalArrangement`, e.g.:

```suggestion
        val arrangement = Arrangement.spacedBy(16.dp)

        if (LocalConfiguration.current.orientation == Configuration.ORIENTATION_PORTRAIT) {
            Column (verticalArrangement = arrangement) {
                NewDomainCard(onNewDomainCardSelected)
                ExistingDomainCard(onExistingDomainCardSelected)
            }
        } else {
            Row (horizontalArrangement = arrangement) {
                NewDomainCard(onNewDomainCardSelected, modifier = Modifier.weight(1f))
                ExistingDomainCard(onExistingDomainCardSelected, modifier = Modifier.weight(1f))
            }
        }
```

justtwago replied:
Wow, `Today I Learned`! Thank you for the suggestion 🤩

mkevins commented on WordPress/src/test/java/org/wordpress/android/ui/domains/management/usedomain/UseDomainViewModelTest.kt:
```diff
@@ -0,0 +1,85 @@
+    @Test
+    fun `WHEN new domain selected THEN track DOMAIN_MANAGEMENT_USE_DOMAIN_SCREEN_NEW_DOMAIN_TAPPED event`() {
+        viewModel.onNewDomainSelected()
+        verify(analyticsTracker).track(DOMAIN_MANAGEMENT_USE_DOMAIN_SCREEN_SHOWN)
```
Thank you for adding tests! Did you perhaps forget to update the verified tracked event in some of these tests? (they're all verifying `DOMAIN_MANAGEMENT_USE_DOMAIN_SCREEN_SHOWN`).

justtwago replied:
Yeah, I copy-pasted the analytics-related tests and forgot to change the events. 🤦‍♂️
 All the tests pass because `DOMAIN_MANAGEMENT_USE_DOMAIN_SCREEN_SHOWN` is initially fired event. Already fixed it. Thank you for catching this!

mkevins reviewed (changes requested):
This is looking great, Artyom! I left a few comments, with a couple of suggestions, and nudge about fixing some of the tests.

Another thing that's been on my mind is whether it'd be worth trying to use material 3, as done in https://github.com/wordpress-mobile/WordPress-Android/pull/19269. Though that PR hasn't landed yet, I think using the theme defined there with all the composables for this feature will help us shape it up for hoisting. That said, I think it's something we could always address later, and consider non-blocking here.

Commit:
Add missing "Use Domain" screen's analytics events to AnalyticsTrackerNosara

Commit:
Fix UseDomainViewModelTest analytics tests

Commit:
UI improvements:
- Use Arrangement for spacing between domain cards.
- Use single-line header for landscape orientation.

justtwago commented:
Thank you @antonis, @paulforgione, @mkevins for all the great suggestions. I addressed all of them and pushed the commits.
<details>
<summary>Updated demo with a single line header in landscape mode:</summary>

https://github.com/wordpress-mobile/WordPress-Android/assets/16563318/038e0be7-9955-4226-8f24-2461246ed3b1

</details>

> Another thing that's been on my mind is whether it'd be worth trying to use material 3, as done in https://github.com/wordpress-mobile/WordPress-Android/pull/19269. Though that PR hasn't landed yet, I think using the theme defined there with all the composables for this feature will help us shape it up for hoisting. That said, I think it's something we could always address later, and consider non-blocking here.

@mkevins, yeah, thanks for reminding me about Material 3. Totally agree we have to stick to it when creating new screens. I thought about this issue, and I'm good with either waiting for your PR to be merged or merging this one first and modifying the screen's theme when Material 3 is available on `trunk` when wiring this screen up to the rest of the flow. Please let me know what would work best here.

antonis reviewed (approved):
Thank you for the changes @justtwago 🙇 
Everything LGTM 🎉

hassaanelgarem commented:
@justtwago FYI: This screen title was updated from "Use Domain" to "Purchase Domain"
Ref: rvVgXupiyONUcJMgPSoMFj-fi-2565:29982#582613784

Commit:
Change usedomain package name to purchasedomain

Commit:
Rename UseDomain keywords to PurchaseDomain for every class and function name

Commit:
Rename USE_DOMAIN keyword to PURCHASE_DOMAIN for every analytics event

Commit:
Rename use_domain keyword to purchase_domain for every string resource

Commit:
Rename Use Domain to Purchase Domain on the screen's toolbar

Commit:
Migrate purchase domain screen to Material 3 theme

justtwago commented:
@mkevins, @antonis, please take a look at the PR one more time. I've made some crucial changes:
* Renamed the screen and all the "Use Domain" keywords in the codebase to "Purchase Domain."
* Migrated the screen to the Material 3-based theme.

PS, I haven't found any strict policy about a branch updating, so after surfing around closed PRs, I noticed that we usually chose `rebase` onto `trunk` instead of `merge` `trunk`, so I did the same. Please let me know if you prefer the `merge` way instead next time.

Commit:
Resolve detekt issues for unused imports and max line length

justtwago commented on WordPress/src/main/java/org/wordpress/android/ui/domains/management/Theme.kt:
```diff
@@ -41,10 +41,12 @@ fun M3ThemeWithoutBackground(
             colorScheme = if (isDarkTheme) {
                 darkColorScheme(
                     error = AppColor.Red30,
+                    surface = AppColor.DarkGray
                 )
             } else {
                 lightColorScheme(
                     error = AppColor.Red50,
+                    surface = AppColor.White
```
I overwrote this because default surface values are incompatible with the rest of the app. Please let me know if you see a better solution for this.

mkevins replied:
This is great, actually. My plan is to migrate the colors, as needed, from [JetpackColors.kt](https://github.com/wordpress-mobile/WordPress-Android/blob/19573b2216b43f5840200378c23dcb0732b86daa/WordPress/src/main/java/org/wordpress/android/ui/compose/theme/JetpackColors.kt), with the best possible semantic mapping. Since there [isn't exactly a one-to-one correspondence between these](https://developer.android.com/jetpack/compose/designsystems/material2-material3#color), migrating them one by one makes sense, imo.

antonis reviewed (approved):
Thank you for the changes @justtwago and for splitting in independent commits that are easy to follow 🙇 
The code changes look good to me and the result matches the design (ref: rvVgXupiyONUcJMgPSoMFj-fi-2565_30081) 🎉

justtwago commented:
I'd love to merge this PR if you don't see any blockers, @mkevins. 
I can't do this with no bypassing branch protections:

<img width="593" alt="Screenshot 2023-10-19 at 12 51 17 PM" src="https://github.com/wordpress-mobile/WordPress-Android/assets/16563318/51e97ff5-5d72-471c-a440-fba032bb1d59">

As for the action buttons' corners radius - I can address it in a follow-up PR when your changes are available on `trunk`.

mkevins reviewed (approved):
This looks good to me! Thanks for updating, and fixing the tests too.

> I'd love to merge this PR if you don't see any blockers

The request changes was just for the tests, which you already fixed.

> I haven't found any strict policy about a branch updating, so after surfing around closed PRs, I noticed that we usually chose rebase onto trunk instead of merge trunk, so I did the same. Please let me know if you prefer the merge way instead next time.

This is totally fine. I think we do a combination of both, but there is one circumstance where merging can make more sense than rebasing: when there or other branches / PRs that are dependent. Then, merging is usually better because the conflict resolution is recorded, and doesn't need to be duplicated.
